docs: code guideline to use mocks vs spies

feloy opened the pull request:
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

add a code guideline to choose between using mocks vs spies, and how to reset/restore them.

### Screenshot / video of UI

N/A

### What issues does this PR fix or reference?

Part of #9821

### How to test this PR?

<!-- Please explain steps to verify the functionality,
do not forget to provide unit/component tests -->

- [ ] Tests are covering the bug fix or the new feature

feloy commented:
> I'm not sure to follow why we would need to do mockRestore
> 
> as myFuncSpy is a new instance at each test run (we recreate the object on each test run)

Some test outside of the `describe('spy on myfunc')` may want to use the original `myfunc`, but as I understand, without `mockRestore`, the function would still be mocked.

benoitf commented:
I tried

```ts
import type { MockInstance} from 'vitest';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

describe('spy on myfunc', () => {
  let myfuncSpy: MockInstance<() => void>;

  beforeEach(() => {
    myfuncSpy = vi.spyOn(console, 'log');
  });

  afterEach(() => {
    // myfuncSpy.mockRestore();
  });

  test('use console.log', () => {
    console.log('hello world');
    expect(myfuncSpy).toHaveBeenCalled();
  });

  test('another one', () => {
    console.warn('warn hello world');
    expect(myfuncSpy).not.toHaveBeenCalled();
  });


  test('third one', () => {
    // not calling so not been called
    expect(myfuncSpy).not.toHaveBeenCalled();
  });



});

```


so no mockRestore and I don't see any issues

feloy commented:
> I tried
> 
> ```ts
> import type { MockInstance} from 'vitest';
> import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
> 
> describe('spy on myfunc', () => {
>   let myfuncSpy: MockInstance<() => void>;
> 
>   beforeEach(() => {
>     myfuncSpy = vi.spyOn(console, 'log');
>   });
> 
>   afterEach(() => {
>     // myfuncSpy.mockRestore();
>   });
> 
>   test('use console.log', () => {
>     console.log('hello world');
>     expect(myfuncSpy).toHaveBeenCalled();
>   });
> 
>   test('another one', () => {
>     console.warn('warn hello world');
>     expect(myfuncSpy).not.toHaveBeenCalled();
>   });
> 
> 
>   test('third one', () => {
>     // not calling so not been called
>     expect(myfuncSpy).not.toHaveBeenCalled();
>   });
> 
> 
> 
> });
> ```
> 
> so no mockRestore and I don't see any issues

Right, it is necessary only when the spy is also mocked, like in:

```ts

describe('spy on myfunc', () => {
  let myfuncSpy: MockInstance<() => void>;

  beforeEach(() => {
    myfuncSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    myfuncSpy.mockRestore();
  });

  test('use console.log', () => {
    console.log('hello world');
    expect(myfuncSpy).toHaveBeenCalled();
  });

  test('another one', () => {
    console.warn('warn hello world');
    expect(myfuncSpy).not.toHaveBeenCalled();
  });


  test('third one', () => {
    // not calling so not been called
    expect(myfuncSpy).not.toHaveBeenCalled();
  });
});

test('real console', () => {
  console.log('==> should be displayed on console');
});
```

In this example, the `real console` does not display anything if `mockRestore` is not called.

I'll change to add an implementation to the spy.

benoitf commented:
but we recommend to use resetAllMocks to have clean state on every test, so still I don't see why we would need mockRestore ?

it'll work as expected:

```
describe('spy on myfunc', () => {
  let myfuncSpy: MockInstance<() => void>;

  beforeEach(() => {
    vi.resetAllMocks();
    myfuncSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
  });

  test('use console.log', () => {
    console.log('hello world');
    expect(myfuncSpy).toHaveBeenCalled();
  });

  test('another one', () => {
    console.warn('warn hello world');
    expect(myfuncSpy).not.toHaveBeenCalled();
  });


  test('third one', () => {
    // not calling so not been called
    expect(myfuncSpy).not.toHaveBeenCalled();
  });
});

test('real console', () => {
  console.log('==> should be displayed on console');
});
```

feloy commented:
> but we recommend to use resetAllMocks to have clean state on every test, so still I don't see why we would need mockRestore ?
> 
> it'll work as expected:
> 
> ```
> describe('spy on myfunc', () => {
>   let myfuncSpy: MockInstance<() => void>;
> 
>   beforeEach(() => {
>     vi.resetAllMocks();
>     myfuncSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
>   });
> 
>   test('use console.log', () => {
>     console.log('hello world');
>     expect(myfuncSpy).toHaveBeenCalled();
>   });
> 
>   test('another one', () => {
>     console.warn('warn hello world');
>     expect(myfuncSpy).not.toHaveBeenCalled();
>   });
> 
> 
>   test('third one', () => {
>     // not calling so not been called
>     expect(myfuncSpy).not.toHaveBeenCalled();
>   });
> });
> 
> test('real console', () => {
>   console.log('==> should be displayed on console');
> });
> ```

No, I cannot see it working, it does not display anything in the console for the `real console` test.

This needs 2 changes:

- use `restoreAllMocks` instead of `resetAllMocks` (1)
- call it from `afterEach`

```ts
describe('spy on myfunc', () => {
  let myfuncSpy: MockInstance<() => void>;

  beforeEach(() => {
    myfuncSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  test('use console.log', () => {
    console.log('hello world');
    expect(myfuncSpy).toHaveBeenCalled();
  });

  test('another one', () => {
    console.warn('warn hello world');
    expect(myfuncSpy).not.toHaveBeenCalled();
  });


  test('third one', () => {
    // not calling so not been called
    expect(myfuncSpy).not.toHaveBeenCalled();
  });
});

test('real console', () => {
  console.log('==> should be displayed on console');
});
```

(1) https://vitest.dev/api/vi#vi-restoreallmocks
```
vi.resetAllMocks
Will call [.mockReset()](https://vitest.dev/api/mock#mockreset) on all spies. This will clear mock history and reset its implementation to an empty function (will return undefined).

vi.restoreAllMocks
Will call [.mockRestore()](https://vitest.dev/api/mock#mockrestore) on all spies. This will clear mock history and reset its implementation to the original one.
```

benoitf commented:
I think originally we were doing

```
  vi.resetAllMocks();
  vi.restoreAllMocks(); 
```

but @dgolovin said we should not do restore as it was part of reset

so now it seems we should have stick to the original rule of doing both ? 🙄

jeffmaury commented on CODE-GUIDELINES.md:
````diff
@@ -37,3 +37,97 @@ test('...', () => {
 });
 ```
+
+### Mock complete modules, spy on parts of module for specific tests
````
Does it apply also to internal modules ?

feloy replied:
I would not like to enforce this point, it would be to the developer to choose the limit of what is mocked/unit tested. I'll develop a little bit more to clarify this point

feloy commented:
I realize that my second example (for spies) is not really realistic, as the spy is created in an object re-created for each test, which makes restoring the spy not necessary. I'll change the example

feloy commented:
> I realize that my second example (for spies) is not really realistic, as the spy is created in an object re-created for each test, which makes restoring the spy not necessary. I'll change the example

I've updated the example, and added some explanations concerning the choice between mock vs spyOn

feloy commented:
@SoniaSandler @benoitf @axel7083 @dgolovin PR ready for review
